### PR TITLE
Plot Editing: Simple & Advanced mode + Undo & Redo

### PR DESCRIPTION
**Pull request by vandenman**
Fixes most of https://github.com/jasp-stats/INTERNAL-jasp/issues/1191
Requires: https://github.com/jasp-stats/jaspGraphs/pull/10

Some things that are not fixed:
- All plots that can't be edited should have a disabled menu entry with a tooltip explaining why it can't be edited.
- Duplicate analysis doesn't copy the state and causes errors.
- Tooltips don't work when a QML thingy is disabled (see the TextField example).

Aside from that, discrete axes needs some work, but that might require a new QML component and/ or different approach.

### CODE REVIEW

File: Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml [hunk 1]
```diff
@@ -81,61 +81,68 @@ Popup
...
+					
+						PlotEditingAxis
 						{
-							id:				axisSeparator
-							height:			1
-							color:			jaspTheme.uiBorder
+							id:				xAxis
+							title:			qsTr("x-axis")
+							axisModel:		plotEditorModel.xAxis
+							visible:		axisDropDown.value === "x-axis"
```
Comment (JorisGoosen) [inline Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml]: It would be better to add an Enum for the axis and select it based on that.

Comment (JorisGoosen) [reply Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml]: To make the selected axis better available here you could add a property to `PlotEditorModel` called `selectedAxis` and have that be a reference to the `PlotEditorAxisModel*` that is current.

But maybe it is better if I implement that.

File: Desktop/results/ploteditoraxismodel.cpp [hunk 2]
```diff
@@ -399,20 +466,22 @@ std::string		AxisModel::LimitsTypeToString(LimitsType type) const
...
-AxisModel::BreaksType	AxisModel::BreaksTypeFromString(std::string type) const
+AxisModel::BreaksType	AxisModel::BreaksTypeFromString(const std::string& type) const
 {
 	if(type == "manual"	)	return BreaksType::BreaksManual;
+	if(type == "NULL")		return BreaksType::BreaksNull;
 							return BreaksType::BreaksRange; //default
 }
 
-AxisModel::LimitsType	AxisModel::LimitsTypeFromString(std::string type) const
+AxisModel::LimitsType	AxisModel::LimitsTypeFromString(const std::string& type) const
```
Comment (JorisGoosen) [inline Desktop/results/ploteditoraxismodel.cpp:484]: Ah yeah, oops

File: Desktop/results/ploteditormodel.cpp [hunk 3]
```diff
@@ -134,6 +143,86 @@ void PlotEditorModel::somethingChanged()
...
+void PlotEditorModel::setUndoEnabled(bool undoEnabled)
+{
+	if (_undoEnabled == undoEnabled)
+		return;
+
+	_undoEnabled = undoEnabled;
+	emit undoEnabledChanged(_undoEnabled);
+}
+
+void PlotEditorModel::setRedoEnabled(bool redoEnabled)
```
Comment (JorisGoosen) [inline Desktop/results/ploteditormodel.cpp]: This whole function isn't necessary and I'll show you later how it can be done more efficiently.

Comment (JorisGoosen) [reply Desktop/results/ploteditormodel.cpp]: (And avoiding double administration, because redo-enabled == redostacksize > 0 )

**Review by JorisGoosen — CHANGES_REQUESTED** [review]
Alright, overall this looks good.
Two things:
- [x] You forgot to add undo/redo buttons to the darkTheme... 
- [x] If you have the "Manual" breaks selected in simple mode and you switch to advanced it selects something else.

Now, I can have a look at that last issue and also clean up the axisselection.
I would like you to add those icons though.

**Comment by JorisGoosen** [comment]
I'll change the axis-enum and the redo enabled later if you like.

**Comment by JorisGoosen** [comment]
(Of course if you do not like I can help you figure how to do those changes)

File: Desktop/mainwindow.cpp [hunk 4]
```diff
@@ -168,7 +168,8 @@ MainWindow::MainWindow(QApplication * application) : QObject(application), _appl
 	qmlRegisterType<JASPDoubleValidator>						("JASP",		1, 0, "JASPDoubleValidator"				);
 	qmlRegisterType<ResultsJsInterface>							("JASP",		1, 0, "ResultsJsInterface"				);
 
-	qmlRegisterUncreatableType<PlotEditor::AxisModel>			("JASP.PlotEditor",		1, 0, "AxisModel",		"Can't make it");
+	qmlRegisterUncreatableType<PlotEditor::AxisModel>			("JASP.PlotEditorAxis",		1, 0, "AxisModel",			"Can't make it");
```
Comment (JorisGoosen) [inline Desktop/mainwindow.cpp]: Isn't it more logical to put both of these QML items inside the same group called "JASP.PlotEditor" ?

That way you can just do `import JASP.PlotEditor 1.0` and then use `AxisModel` and `PlotEditorModel` directly.
The way you've done it now is rather confusing.

Comment (vandenman) [reply Desktop/mainwindow.cpp]: that somehow didn't work before, but I'll try again :)

Comment (vandenman) [reply Desktop/mainwindow.cpp]: and now it does work ...

File: Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml (hunk 1 above)
Comment (vandenman) [reply Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml]: I changed it, let me know if you have any further comments!

**Review by JorisGoosen — APPROVED** [review]
Ok, this is starting to look good.
I will be merging it so other people can play around with it.

But there are some improvements we can (and should) still make.

**Pull request merged**